File: libs/shared-models/src/alert.ts
import { BoundingBox, Coordinate, PaginationRequest, PaginationResponse } from './common';
import { EntityType } from './entity';

// Canonical definitions in @sentinel/proto-gen (proto/alert.proto)
export enum AlertSeverity {
  INFO = 'INFO',
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL',
}

export enum AlertType {
  GEOFENCE_BREACH = 'GEOFENCE_BREACH',
  SPEED_ANOMALY = 'SPEED_ANOMALY',
  PATTERN_DEVIATION = 'PATTERN_DEVIATION',
  PROXIMITY = 'PROXIMITY',
  COMMUNICATION_ANOMALY = 'COMMUNICATION_ANOMALY',
  NEW_ENTITY = 'NEW_ENTITY',
  ENTITY_CLASSIFICATION_CHANGE = 'ENTITY_CLASSIFICATION_CHANGE',
  GEOFENCE_ENTRY = 'GEOFENCE_ENTRY',
  GEOFENCE_EXIT = 'GEOFENCE_EXIT',
  PATTERN_MATCH = 'PATTERN_MATCH',
  CUSTOM = 'CUSTOM',
  PROXIMITY_ALERT = 'PROXIMITY_ALERT',
  ENTITY_LOST = 'ENTITY_LOST',
  PATTERN_DETECTED = 'PATTERN_DETECTED',
  LINK_CHANGE = 'LINK_CHANGE',
  CLASSIFICATION_CHANGE = 'CLASSIFICATION_CHANGE',
  SYSTEM = 'SYSTEM',
}

export enum RuleType {
  GEOFENCE = 'GEOFENCE',
  SPEED_THRESHOLD = 'SPEED_THRESHOLD',
  PROXIMITY = 'PROXIMITY',
  PATTERN = 'PATTERN',
}

export interface GeofenceRule {
  ruleId: string;
  name: string;
  polygon: Coordinate[];
  monitoredEntityTypes: EntityType[];
  triggerOnEnter: boolean;
  triggerOnExit: boolean;
}

export interface Alert {
  alertId: string;
  alertType: AlertType;
  severity: AlertSeverity;
  title: string;
  description: string;
  entityId: string;
  relatedEntityIds: string[];
  position?: Coordinate;
  ruleId?: string;
  metadata: Record<string, string>;
  createdAt: string;
  acknowledgedAt?: string;
  acknowledgedBy?: string;
  resolvedAt?: string;
}

export interface AlertRule {
  id: string;
  name: string;
  ruleType: AlertType;
  config: Record<string, unknown>;
  monitoredEntityTypes: EntityType[];
  severity: AlertSeverity;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export enum AlertEventType {
  TRIGGERED = 'TRIGGERED',
  ACKNOWLEDGED = 'ACKNOWLEDGED',
  RESOLVED = 'RESOLVED',
}

export interface AlertEvent {
  alert: Alert;
  eventType: AlertEventType;
}

export interface GetAlertsRequest extends PaginationRequest {
  severityMin?: AlertSeverity;
  alertTypes?: AlertType[];
  entityId?: string;
  boundingBox?: BoundingBox;
  startTime?: string;
  endTime?: string;
  acknowledged?: boolean;
}

export interface GetAlertsResponse {
  alerts: Alert[];
  pagination: PaginationResponse;
}

File: libs/common-go/models/alert.go
package models

// AlertSeverity constants.
const (
	AlertSeverityInfo     = "INFO"
	AlertSeverityLow      = "LOW"
	AlertSeverityMedium   = "MEDIUM"
	AlertSeverityHigh     = "HIGH"
	AlertSeverityCritical = "CRITICAL"
)

// AlertType constants.
const (
	AlertTypeGeofenceBreach           = "GEOFENCE_BREACH"
	AlertTypeSpeedAnomaly             = "SPEED_ANOMALY"
	AlertTypePatternDeviation         = "PATTERN_DEVIATION"
	AlertTypeProximity                = "PROXIMITY"
	AlertTypeNewEntity                = "NEW_ENTITY"
	AlertTypeGeofenceEntry            = "GEOFENCE_ENTRY"
	AlertTypeGeofenceExit             = "GEOFENCE_EXIT"
	AlertTypePatternMatch             = "PATTERN_MATCH"
	AlertTypeCustom                   = "CUSTOM"
	AlertTypeEntityLost               = "ENTITY_LOST"
	AlertTypePatternDetected          = "PATTERN_DETECTED"
	AlertTypeLinkChange               = "LINK_CHANGE"
	AlertTypeClassificationChange     = "CLASSIFICATION_CHANGE"
	AlertTypeSystem                   = "SYSTEM"
)

// RuleType constants.
const (
	RuleTypeGeofence       = "GEOFENCE"
	RuleTypeSpeedThreshold = "SPEED_THRESHOLD"
	RuleTypeProximity      = "PROXIMITY"
	RuleTypePattern        = "PATTERN"
)

// Alert represents an alert record.
type Alert struct {
	AlertID          string                 `json:"alertId"`
	AlertType        string                 `json:"alertType"`
	Severity         string                 `json:"severity"`
	Title            string                 `json:"title"`
	Description      string                 `json:"description"`
	EntityID         string                 `json:"entityId"`
	RelatedEntityIDs []string               `json:"relatedEntityIds"`
	Position         *Coordinate            `json:"position,omitempty"`
	RuleID           *string                `json:"ruleId,omitempty"`
	Metadata         map[string]string      `json:"metadata"`
	CreatedAt        string                 `json:"createdAt"`
	AcknowledgedAt   *string                `json:"acknowledgedAt,omitempty"`
	AcknowledgedBy   *string                `json:"acknowledgedBy,omitempty"`
	ResolvedAt       *string                `json:"resolvedAt,omitempty"`
}

// AlertRule represents an alert rule configuration.
type AlertRule struct {
	ID                   string                 `json:"id"`
	Name                 string                 `json:"name"`
	RuleType             string                 `json:"ruleType"`
	Config               map[string]interface{} `json:"config"`
	MonitoredEntityTypes []string               `json:"monitoredEntityTypes"`
	Severity             string                 `json:"severity"`
	Enabled              bool                   `json:"enabled"`
	CreatedAt            string                 `json:"createdAt"`
	UpdatedAt            string                 `json:"updatedAt"`
}
